# E-Commerce Customer Churn & Retention Analysis

## 1. Data Understanding

The objective of this section is to understand the structure, variables, and overall quality of the e-commerce customer dataset before performing any cleaning or analysis.


### Import Required Library

Pandas is imported for loading, inspecting, and analyzing the dataset.


In [1]:
import pandas as pd


### Load the Dataset

The customer-level data is loaded from the `E Comm` worksheet of the original Excel workbook.


In [2]:
df = pd.read_excel('../Data/E Commerce Dataset.xlsx', sheet_name="E Comm")


### Dataset Preview

The first few records are reviewed to confirm that the dataset was loaded correctly and to obtain an initial view of its structure and variables.


In [3]:
df.head()


,CustomerID,Churn,Tenure,PreferredLoginDevice,CityTier,WarehouseToHome,PreferredPaymentMode,Gender,HourSpendOnApp,NumberOfDeviceRegistered,PreferedOrderCat,SatisfactionScore,MaritalStatus,NumberOfAddress,Complain,OrderAmountHikeFromlastYear,CouponUsed,OrderCount,DaySinceLastOrder,CashbackAmount
0,50001,1,4.0,Mobile Phone,3,6.0,Debit Card,Female,3.0,3,Laptop & Accessory,2,Single,9,1,11.0,1.0,1.0,5.0,159.93
1,50002,1,NaN,Phone,1,8.0,UPI,Male,3.0,4,Mobile,3,Single,7,1,15.0,0.0,1.0,0.0,120.90
2,50003,1,NaN,Phone,1,30.0,Debit Card,Male,2.0,4,Mobile,3,Single,6,1,14.0,0.0,1.0,3.0,120.28
3,50004,1,0.0,Phone,3,15.0,Debit Card,Male,2.0,4,Laptop & Accessory,5,Single,8,0,23.0,0.0,1.0,3.0,134.07
4,50005,1,0.0,Phone,1,12.0,CC,Male,NaN,3,Mobile,5,Single,3,0,11.0,1.0,1.0,3.0,129.60


### Dataset Structure

The dataset dimensions and column names are inspected to understand the number of customer records and variables available for analysis.


In [4]:
df.shape


(5630, 20)

The dataset contains **5,630 customer records and 20 features** describing customer demographics, engagement, purchasing behavior, service experience, and churn status.


In [5]:
df.columns


Index(['CustomerID', 'Churn', 'Tenure', 'PreferredLoginDevice', 'CityTier',
       'WarehouseToHome', 'PreferredPaymentMode', 'Gender', 'HourSpendOnApp',
       'NumberOfDeviceRegistered', 'PreferedOrderCat', 'SatisfactionScore',
       'MaritalStatus', 'NumberOfAddress', 'Complain',
       'OrderAmountHikeFromlastYear', 'CouponUsed', 'OrderCount',
       'DaySinceLastOrder', 'CashbackAmount'],
      dtype='str')

### Data Types and Completeness

The dataset structure is examined to identify the data type of each variable and obtain an initial view of data completeness.

Columns with fewer than 5,630 non-null observations indicate the presence of missing values that require further assessment.


In [6]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   CustomerID                   5630 non-null   int64  
 1   Churn                        5630 non-null   int64  
 2   Tenure                       5366 non-null   float64
 3   PreferredLoginDevice         5630 non-null   str    
 4   CityTier                     5630 non-null   int64  
 5   WarehouseToHome              5379 non-null   float64
 6   PreferredPaymentMode         5630 non-null   str    
 7   Gender                       5630 non-null   str    
 8   HourSpendOnApp               5375 non-null   float64
 9   NumberOfDeviceRegistered     5630 non-null   int64  
 10  PreferedOrderCat             5630 non-null   str    
 11  SatisfactionScore            5630 non-null   int64  
 12  MaritalStatus                5630 non-null   str    
 13  NumberOfAddress              

### Missing Value Assessment

Missing values are quantified by both count and percentage to understand the extent of incomplete data across the dataset.


In [7]:
df.isnull().sum()


CustomerID                       0
Churn                            0
Tenure                         264
PreferredLoginDevice             0
CityTier                         0
WarehouseToHome                251
PreferredPaymentMode             0
Gender                           0
HourSpendOnApp                 255
NumberOfDeviceRegistered         0
PreferedOrderCat                 0
SatisfactionScore                0
MaritalStatus                    0
NumberOfAddress                  0
Complain                         0
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
CashbackAmount                   0
dtype: int64

In [8]:
((df.isnull().sum() / len(df)) * 100).round(2)


CustomerID                     0.00
Churn                          0.00
Tenure                         4.69
PreferredLoginDevice           0.00
CityTier                       0.00
WarehouseToHome                4.46
PreferredPaymentMode           0.00
Gender                         0.00
HourSpendOnApp                 4.53
NumberOfDeviceRegistered       0.00
PreferedOrderCat               0.00
SatisfactionScore              0.00
MaritalStatus                  0.00
NumberOfAddress                0.00
Complain                       0.00
OrderAmountHikeFromlastYear    4.71
CouponUsed                     4.55
OrderCount                     4.58
DaySinceLastOrder              5.45
CashbackAmount                 0.00
dtype: float64

Missing values are present in seven numerical columns:

- `Tenure`
- `WarehouseToHome`
- `HourSpendOnApp`
- `OrderAmountHikeFromlastYear`
- `CouponUsed`
- `OrderCount`
- `DaySinceLastOrder`

Each affected column contains approximately **4.5%–5.5% missing observations**. No treatment is performed at this stage; the appropriate approach will be determined during data cleaning.


### Duplicate Assessment

The dataset is checked for both completely duplicated records and duplicated `CustomerID` values.

Because `CustomerID` is expected to uniquely identify each customer, checking its uniqueness helps verify that individual customers are not represented multiple times.


In [9]:
df.duplicated().sum()


np.int64(0)

In [10]:
df['CustomerID'].duplicated().sum()


np.int64(0)

No fully duplicated records or duplicated CustomerIDs were identified. Therefore, each of the **5,630 records represents a unique customer**.


### Numerical Data Assessment

Descriptive statistics are reviewed primarily as a data-quality check to examine the ranges and general characteristics of the numerical variables.

Minimum and maximum values are inspected to identify potentially unusual or implausible observations that may require further investigation.


In [11]:
df.describe()


,CustomerID,Churn,Tenure,CityTier,WarehouseToHome,HourSpendOnApp,NumberOfDeviceRegistered,SatisfactionScore,NumberOfAddress,Complain,OrderAmountHikeFromlastYear,CouponUsed,OrderCount,DaySinceLastOrder,CashbackAmount
count,5630.000000,5630.000000,5366.000000,5630.000000,5379.000000,5375.000000,5630.000000,5630.000000,5630.000000,5630.000000,5365.000000,5374.000000,5372.000000,5323.000000,5630.000000
mean,52815.500000,0.168384,10.189899,1.654707,15.639896,2.931535,3.688988,3.066785,4.214032,0.284902,15.707922,1.751023,3.008004,4.543491,177.223030
std,1625.385339,0.374240,8.557241,0.915389,8.531475,0.721926,1.023999,1.380194,2.583586,0.451408,3.675485,1.894621,2.939680,3.654433,49.207036
min,50001.000000,0.000000,0.000000,1.000000,5.000000,0.000000,1.000000,1.000000,1.000000,0.000000,11.000000,0.000000,1.000000,0.000000,0.000000
25%,51408.250000,0.000000,2.000000,1.000000,9.000000,2.000000,3.000000,2.000000,2.000000,0.000000,13.000000,1.000000,1.000000,2.000000,145.770000
50%,52815.500000,0.000000,9.000000,1.000000,14.000000,3.000000,4.000000,3.000000,3.000000,0.000000,15.000000,1.000000,2.000000,3.000000,163.280000
75%,54222.750000,0.000000,16.000000,3.000000,20.000000,3.000000,4.000000,4.000000,6.000000,1.000000,18.000000,2.000000,3.000000,7.000000,196.392500
max,55630.000000,1.000000,61.000000,3.000000,127.000000,5.000000,6.000000,5.000000,22.000000,1.000000,26.000000,16.000000,16.000000,46.000000,324.990000


### Investigation of Unusually High Values

Several numerical variables show relatively high maximum values compared with their typical ranges. Their highest observations are reviewed before deciding whether they represent data-quality issues.

Extreme observations are not automatically treated as errors simply because they are unusually high.


In [12]:
columns = [
    'Tenure',
    'WarehouseToHome',
    'NumberOfAddress',
    'CouponUsed',
    'DaySinceLastOrder',
    'OrderCount'
]

for col in columns:
    print(f"\n{col}")
    print(df[col].sort_values(ascending=False).head(10).values)



Tenure
[61. 60. 51. 50. 31. 31. 31. 31. 31. 31.]

WarehouseToHome
[127. 126.  36.  36.  36.  36.  36.  36.  36.  36.]

NumberOfAddress
[22 21 20 19 11 11 11 11 11 11]

CouponUsed
[16. 16. 15. 14. 14. 14. 14. 14. 13. 13.]

DaySinceLastOrder
[46. 31. 30. 18. 18. 18. 18. 18. 18. 18.]

OrderCount
[16. 16. 16. 16. 16. 16. 16. 16. 16. 16.]


The upper-range observations were inspected and there was insufficient evidence to classify them as invalid. These values will therefore be retained rather than removed solely because they are unusually high.


### Categorical Data Assessment

Categorical variables are inspected to identify inconsistent labels, duplicate representations of the same category, and potentially invalid values.


In [13]:
categoricalColumns = [
    'PreferredLoginDevice',
    'PreferredPaymentMode',
    'Gender',
    'PreferedOrderCat',
    'MaritalStatus'
]

for col in categoricalColumns:
    print(f"\n{col}")
    print(df[col].unique())



PreferredLoginDevice
<StringArray>
['Mobile Phone', 'Phone', 'Computer']
Length: 3, dtype: str

PreferredPaymentMode
<StringArray>
['Debit Card', 'UPI', 'CC', 'Cash on Delivery', 'E wallet', 'COD',
 'Credit Card']
Length: 7, dtype: str

Gender
<StringArray>
['Female', 'Male']
Length: 2, dtype: str

PreferedOrderCat
<StringArray>
['Laptop & Accessory', 'Mobile', 'Mobile Phone', 'Others', 'Fashion',
 'Grocery']
Length: 6, dtype: str

MaritalStatus
<StringArray>
['Single', 'Divorced', 'Married']
Length: 3, dtype: str


Three categorical variables contain inconsistent labels representing the same underlying categories:

- `PreferredLoginDevice`: `Phone` and `Mobile Phone`
- `PreferedOrderCat`: `Mobile` and `Mobile Phone`
- `PreferredPaymentMode`: `CC` and `Credit Card`; `COD` and `Cash on Delivery`

These inconsistencies will be standardized during the data-cleaning stage.


## Data Quality Findings

The initial data-understanding process identified the following findings:

1. The dataset contains **5,630 customer records and 20 variables**.
2. Each record represents a unique customer, with **no duplicate rows or duplicate CustomerIDs**.
3. **Seven numerical columns contain missing values**, with approximately 4.5%–5.5% missing observations in each affected column.
4. Numerical ranges were reviewed for potentially unusual observations. The investigated upper-range values were retained because there was insufficient evidence to classify them as invalid.
5. **Three categorical columns contain inconsistent category labels** that require standardization:
   - `PreferredLoginDevice`
   - `PreferedOrderCat`
   - `PreferredPaymentMode`
6. No transformations or cleaning were performed during this stage. The identified data-quality issues will be addressed separately during the data-cleaning process.

### Next Step

The next stage will focus on cleaning and preparing the dataset by:

- standardizing inconsistent categorical labels;
- determining an appropriate treatment for missing values;
- validating the cleaned dataset before beginning exploratory and churn analysis.
